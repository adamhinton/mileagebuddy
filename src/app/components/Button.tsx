// _______________________________________________________
// This is a generic button component for forms with optional confirmation dialog
// It can be used for submit, clear form, cancel, or any other form actions
// The confirmation dialog is optional and configurable
// _______________________________________________________

// TODO write tests for this

import { ButtonHTMLAttributes, useState } from "react";
import ConfirmationDialog from "./ConfirmationDialog";

type ButtonVariant = "primary" | "danger" | "secondary";

/**
 * Configuration options for the confirmation dialog. All fields optional because they have default values
 */
type ConfirmationDialogOptions = {
	title?: string;
	message?: string;
	confirmButtonText?: string;
	cancelButtonText?: string;
};

/**
 * Base props shared by all button configurations
 *
 * Don't use this type. It's extended to make different config options. Use ButtonProps instead. This type would allow invalid states.
 */
type ButtonPropsBase = {
	// Don't use this type - see above
	onClick: () => void;
	/** Button text */
	text: string;
	variant?: ButtonVariant;
	isDisabled?: boolean;
	className?: string;
	ariaLabel?: string;
	type?: ButtonHTMLAttributes<HTMLButtonElement>["type"];
};

/**
 * Props for button with confirmation dialog required
 */
type ButtonPropsWithConfirmationDialog = ButtonPropsBase & {
	isConfirmationRequired: true;
	confirmationDialogOptions: ConfirmationDialogOptions;
};

/**
 * Props for button without confirmation dialog
 */
type ButtonPropsWithoutConfirmationDialog = ButtonPropsBase & {
	isConfirmationRequired: false;
	confirmationDialogOptions?: never;
};

/**
 * Union type representing all valid button configurations
 *
 * TS types should only/always represent valid states. So, if isConfirmationRequired is true, confirmationDialogOptions must be provided. If it's false, confirmationDialogOptions must not be provided.
 */
export type ButtonProps =
	| ButtonPropsWithConfirmationDialog
	| ButtonPropsWithoutConfirmationDialog;

/**
 * Generic form button component that can optionally show a confirmation dialog
 *
 * Can be used for submit, clear, cancel, or any other form actions
 */
const Button = ({
	onClick,
	text,
	isConfirmationRequired = false,
	// TODO not sure if this will make conf dialog show up when that's not intended
	confirmationDialogOptions = {
		title: "Confirm Action",
		message: "Are you sure you want to proceed? This action cannot be undone.",
		confirmButtonText: "Confirm",
		cancelButtonText: "Cancel",
	},
	variant = "primary",
	isDisabled = false,
	className = "",
	ariaLabel,
	type = "button",
}: ButtonProps) => {
	// Show confirmation dialog when user clicks the button
	const [showConfirmation, setShowConfirmation] = useState(false);

	const handleClick = () => {
		if (isConfirmationRequired) {
			setShowConfirmation(true);
		} else {
			onClick();
		}
	};

	// When user confirms the action
	const handleConfirm = () => {
		onClick();
		setShowConfirmation(false);
	};

	// When user cancels the action
	const handleCancel = () => {
		setShowConfirmation(false);
	};

	// Determine button style based on variant
	const buttonStyles = {
		primary:
			"bg-primary text-white hover:bg-primary-600 focus:ring-primary-200",
		danger: "bg-red-600 text-white hover:bg-red-700 focus:ring-red-200",
		secondary:
			"bg-background-elevated text-neutral-text border border-primary-100 hover:bg-background-highlight focus:ring-primary-200 dark:border-primary-200",
	};

	return (
		<>
			<button
				type={type || "button"}
				onClick={handleClick}
				className={`px-4 py-2 rounded-md transition-colors focus:outline-none focus:ring-2 
          text-sm sm:text-base ${buttonStyles[variant]} ${className}`}
				aria-label={ariaLabel || text}
				disabled={isDisabled}
			>
				{text}
			</button>

			{isConfirmationRequired && (
				<ConfirmationDialog
					title={confirmationDialogOptions.title || "Confirm Action"}
					message={
						confirmationDialogOptions.message ||
						"Are you sure you want to proceed?"
					}
					confirmButtonText={
						confirmationDialogOptions.confirmButtonText || "Confirm"
					}
					cancelButtonText={
						confirmationDialogOptions.cancelButtonText || "Cancel"
					}
					onConfirm={handleConfirm}
					onCancel={handleCancel}
					isOpen={showConfirmation}
				/>
			)}
		</>
	);
};

export default Button;
