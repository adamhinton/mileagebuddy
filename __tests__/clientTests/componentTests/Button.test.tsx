import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Button from "@/app/components/Button";

const mockOnClick = jest.fn();

const testConfirmationDialogOptions = {
	title: "Test Confirmation",
	message: "Are you sure?",
	confirmButtonText: "Yes",
	cancelButtonText: "No",
};

describe("Button.tsx", () => {
	describe("Basic rendering tests", () => {
		it("renders without errors when confirmation isn't required", () => {
			render(
				<Button
					text="Test Button"
					onClick={mockOnClick}
					isConfirmationRequired={false}
				/>
			);

			const button = screen.getByRole("button", { name: /test button/i });
			expect(button).toBeInTheDocument();
		});

		it("renders without errors when confirmation is required", () => {
			render(
				<Button
					text="Test Button"
					onClick={mockOnClick}
					isConfirmationRequired={true}
					confirmationDialogOptions={testConfirmationDialogOptions}
				/>
			);

			const button = screen.getByRole("button", { name: /test button/i });
			expect(button).toBeInTheDocument();
		});

		it("displays the provided text", () => {
			const buttonText = "Click Me";
			render(
				<Button
					text={buttonText}
					onClick={mockOnClick}
					isConfirmationRequired={false}
				/>
			);

			const button = screen.getByRole("button", { name: /click me/i });
			expect(button).toHaveTextContent(buttonText);
		});
	});

	describe("Button variants", () => {
		it("renders primary variant without errors", () => {
			render(
				<Button
					text="Primary Button"
					onClick={mockOnClick}
					variant="primary"
					isConfirmationRequired={false}
				/>
			);

			const button = screen.getByRole("button", { name: /primary button/i });
			expect(button).toHaveClass("bg-primary");
			expect(button).toHaveClass("text-white");
		});

		it("renders danger variant without errors", () => {
			render(
				<Button
					text="Danger Button"
					onClick={mockOnClick}
					variant="danger"
					isConfirmationRequired={false}
				/>
			);

			const button = screen.getByRole("button", { name: /danger button/i });
			expect(button).toHaveClass("bg-red-600");
			expect(button).toHaveClass("text-white");
		});

		it("renders secondary variant without errors", () => {
			render(
				<Button
					text="Secondary Button"
					onClick={mockOnClick}
					variant="secondary"
					isConfirmationRequired={false}
				/>
			);

			const button = screen.getByRole("button", { name: /secondary button/i });
			expect(button).toHaveClass("bg-background-elevated");
			expect(button).toHaveClass("text-neutral-text");
			expect(button).toHaveClass("border-primary-100");
		});
	});

	describe("Button functionality", () => {
		it("calls onClick handler when clicked", () => {
			render(
				<Button
					text="Click Me"
					onClick={mockOnClick}
					isConfirmationRequired={false}
				/>
			);

			const button = screen.getByRole("button", {
				name: /click me/i,
			});
			fireEvent.click(button);

			expect(mockOnClick).toHaveBeenCalledTimes(1);
		});

		it("is disabled when isDisabled prop is true", () => {
			render(
				<Button
					text="Disabled Button"
					onClick={mockOnClick}
					isDisabled={true}
					isConfirmationRequired={false}
				/>
			);

			const button = screen.getByRole("button", {
				name: /disabled button/i,
			});
			expect(button).toBeDisabled();

			// Verify clicking doesn't trigger the onClick handler when disabled
			fireEvent.click(button);
			expect(mockOnClick).not.toHaveBeenCalled();
		});

		it("applies custom className when provided", () => {
			const customClass = "test-custom-class";
			render(
				<Button
					text="Custom Class"
					onClick={mockOnClick}
					className={customClass}
					isConfirmationRequired={false}
				/>
			);

			const button = screen.getByRole("button", {
				name: /custom class/i,
			});
			expect(button).toHaveClass(customClass);
		});

		it("uses provided type attribute", () => {
			render(
				<Button
					text="Submit Button"
					onClick={mockOnClick}
					type="submit"
					isConfirmationRequired={false}
				/>
			);

			const button = screen.getByRole("button", {
				name: /submit button/i,
			});
			expect(button).toHaveAttribute("type", "submit");
		});

		it("uses provided aria-label or falls back to text", () => {
			// With custom aria-label
			render(
				<Button
					text="Button Text"
					onClick={mockOnClick}
					ariaLabel="Custom Aria Label"
					isConfirmationRequired={false}
				/>
			);
			const buttonWithAriaLabel = screen.getByRole("button", {
				name: /custom aria label/i,
			});
			expect(buttonWithAriaLabel).toHaveAttribute(
				"aria-label",
				"Custom Aria Label"
			);

			// Fallback to text when no aria-label provided
			render(
				<Button
					text="Fallback Text"
					onClick={mockOnClick}
					isConfirmationRequired={false}
				/>
			);
			const buttonWithoutAriaLabel = screen.getByRole("button", {
				name: /fallback text/i,
			});
			expect(buttonWithoutAriaLabel).toHaveAttribute(
				"aria-label",
				"Fallback Text"
			);
		});
	});

	// The actual confirmation dialog component is tested in ConfirmationDialog.test.tsx
	describe("Confirmation dialog", () => {
		it("shows confirmation dialog when isConfirmationRequired is true", () => {
			render(
				<Button
					text="Delete Item"
					onClick={mockOnClick}
					isConfirmationRequired={true}
					confirmationDialogOptions={{
						title: "Confirm Delete",
						message: "Are you sure you want to delete this item?",
						confirmButtonText: "Delete",
						cancelButtonText: "Cancel",
					}}
				/>
			);

			const button = screen.getByRole("button", { name: /delete item/i });
			fireEvent.click(button);

			expect(screen.getByText("Confirm Delete")).toBeInTheDocument();
			expect(
				screen.getByText("Are you sure you want to delete this item?")
			).toBeInTheDocument();

			expect(mockOnClick).not.toHaveBeenCalled();
		});

		it("does not show confirmation dialog when isConfirmationRequired is false", () => {
			render(
				<Button
					text="Save Item"
					onClick={mockOnClick}
					isConfirmationRequired={false}
				/>
			);

			const button = screen.getByRole("button", { name: /save item/i });
			fireEvent.click(button);

			expect(mockOnClick).toHaveBeenCalledTimes(1);

			// "Confirm Action" is the default text for the confirmation dialog title
			expect(screen.queryByText("Confirm Action")).not.toBeInTheDocument();
		});

		it("calls onClick handler after confirming in dialog", () => {
			render(
				<Button
					text="Delete Item"
					onClick={mockOnClick}
					isConfirmationRequired={true}
					confirmationDialogOptions={{
						title: "Confirm Delete",
						message: "Are you sure?",
						confirmButtonText: "Yes, Delete",
						cancelButtonText: "No, Cancel",
					}}
				/>
			);

			const button = screen.getByRole("button", { name: /delete item/i });
			fireEvent.click(button);

			const confirmButton = screen.getByRole("button", {
				name: /yes, delete/i,
			});

			expect(mockOnClick).not.toHaveBeenCalled();

			fireEvent.click(confirmButton);

			expect(mockOnClick).toHaveBeenCalledTimes(1);
		});

		it("does not call onClick handler when canceling in dialog", () => {
			render(
				<Button
					text="Delete Item"
					onClick={mockOnClick}
					isConfirmationRequired={true}
					confirmationDialogOptions={{
						title: "Confirm Delete",
						message: "Are you sure?",
						confirmButtonText: "Yes, Delete",
						cancelButtonText: "No, Cancel",
					}}
				/>
			);

			const button = screen.getByRole("button", { name: /delete item/i });
			fireEvent.click(button);

			const cancelButton = screen.getByRole("button", {
				name: /no, cancel/i,
			});
			fireEvent.click(cancelButton);

			expect(mockOnClick).not.toHaveBeenCalled();
		});

		it("shows custom dialog text when confirmationDialogOptions are provided", () => {
			const customTitle = "Custom Dialog Title";
			const customMessage = "This is a custom confirmation message";
			const customConfirmText = "Proceed";
			const customCancelText = "Go Back";

			render(
				<Button
					text="Action Button"
					onClick={mockOnClick}
					isConfirmationRequired={true}
					confirmationDialogOptions={{
						title: customTitle,
						message: customMessage,
						confirmButtonText: customConfirmText,
						cancelButtonText: customCancelText,
					}}
				/>
			);

			const button = screen.getByRole("button", { name: /action button/i });
			fireEvent.click(button);

			expect(screen.getByText(customTitle)).toBeInTheDocument();
			expect(screen.getByText(customMessage)).toBeInTheDocument();
			expect(screen.getByText(customConfirmText)).toBeInTheDocument();
			expect(screen.getByText(customCancelText)).toBeInTheDocument();
		});
	});
});
