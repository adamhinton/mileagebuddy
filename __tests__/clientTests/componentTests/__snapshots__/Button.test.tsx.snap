// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.tsx Snapshot tests matches snapshot for button with confirmation dialog open from 4.30.2025 1`] = `
<div>
  <button
    aria-label="Delete Item"
    class="px-4 py-2 rounded-md transition-colors focus:outline-none focus:ring-2 
          text-sm sm:text-base bg-primary text-white hover:bg-primary-600 focus:ring-primary-200 "
    type="button"
  >
    Delete Item
  </button>
  <div
    class="fixed inset-0 bg-black bg-opacity-30 flex items-center justify-center z-50"
  >
    <div
      class="bg-white dark:bg-gray-800 p-6 rounded-md shadow-lg max-w-sm w-full"
    >
      <h3
        class="text-lg font-medium mb-3"
      >
        Confirm Delete
      </h3>
      <div
        class="mb-4"
      >
        Are you sure you want to delete this item?
      </div>
      <div
        class="flex justify-end space-x-3"
      >
        <button
          class="px-4 py-2 rounded-md border border-gray-300 hover:bg-gray-100 dark:border-gray-600 dark:hover:bg-gray-700"
          type="button"
        >
          Cancel
        </button>
        <button
          class="px-4 py-2 rounded-md bg-red-600 text-white hover:bg-red-700"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Button.tsx Snapshot tests matches snapshot for danger button from 4.30.2025 1`] = `
<div>
  <button
    aria-label="Danger Button"
    class="px-4 py-2 rounded-md transition-colors focus:outline-none focus:ring-2 
          text-sm sm:text-base bg-red-600 text-white hover:bg-red-700 focus:ring-red-200 "
    type="button"
  >
    Danger Button
  </button>
</div>
`;

exports[`Button.tsx Snapshot tests matches snapshot for disabled button from 4.30.2025 1`] = `
<div>
  <button
    aria-label="Disabled Button"
    class="px-4 py-2 rounded-md transition-colors focus:outline-none focus:ring-2 
          text-sm sm:text-base bg-primary text-white hover:bg-primary-600 focus:ring-primary-200 "
    disabled=""
    type="button"
  >
    Disabled Button
  </button>
</div>
`;

exports[`Button.tsx Snapshot tests matches snapshot for primary button from 4.30.2025 1`] = `
<div>
  <button
    aria-label="Primary Button"
    class="px-4 py-2 rounded-md transition-colors focus:outline-none focus:ring-2 
          text-sm sm:text-base bg-primary text-white hover:bg-primary-600 focus:ring-primary-200 "
    type="button"
  >
    Primary Button
  </button>
</div>
`;

exports[`Button.tsx Snapshot tests matches snapshot for secondary button from 4.30.2025 1`] = `
<div>
  <button
    aria-label="Secondary Button"
    class="px-4 py-2 rounded-md transition-colors focus:outline-none focus:ring-2 
          text-sm sm:text-base bg-background-elevated text-neutral-text border border-primary-100 hover:bg-background-highlight focus:ring-primary-200 dark:border-primary-200 "
    type="button"
  >
    Secondary Button
  </button>
</div>
`;
